### ISSUE TITLE
Problem when upgrading cluster 4.1.8 -> 4.2.3

### ISSUE THREAD

**Issue by shm-dmitry** [crate/crate#10415]
**CrateDB version**: 4.1.8

**Problem description**:
I have cluster with 5 nodes (crate 4.1.8).
I trying to upgrade cluster to crate 4.2.3 using rolling upgrade.
I stopped one node, upgraded binaries and start it.
It joined into cluster.
But I started to receive errors on some SQL requests.
This sample successfully executes on crate 4.1.8, and on crate 4.2.3, but failed when cluster have mixed nodes: 4 nodes with 4.1.8 and 1 node with 4.2.3.

```
SQLActionException[NotSerializableExceptionWrapper: unhandled_server_exception: java.lang.NullPointerException]
```
Also I seen next error in logs, but I dont know how to reproduce it.
```
ERROR: NotSerializableExceptionWrapper: unhandled_server_exception: java.lang.UnsupportedOperationException: Unknown function: MATCH((geo.shape), {"coordinates"=[37.825565, 55.563923], "type"='Point'}) USING intersects, no overload found for matching argument types: (object, geo_shape, text, object). Possible candidates: match(object, text, text, object):boolean
```
```
ERROR: SQLParseException: Couldn't create executionContexts from
NodeOperations: [2=>NodeOp{ phase{id=2/distributed merge, nodes=[RTq8D5JqRWaTyFALEZIwkg, 72yRyTOaQu6pZJ1ErWvkpQ, oK0gS38tQIa20YqftdLPgw], dist=BROADCAST}, downstreamNodes=[oK0gS38tQIa20YqftdLPgw], downstreamPhase=3, downstreamInputId=0}, 1=>NodeOp{ phase{id=1/distWindowAgg, nodes=[RTq8D5JqRWaTyFALEZIwkg, 72yRyTOaQu6pZJ1ErWvkpQ, oK0gS38tQIa20YqftdLPgw], dist=MODULO}, downstreamNodes=[72yRyTOaQu6pZJ1ErWvkpQ, RTq8D5JqRWaTyFALEZIwkg, oK0gS38tQIa20YqftdLPgw], downstreamPhase=2, downstreamInputId=0}, 0=>NodeOp{ phase{id=0/collect, nodes=[72yRyTOaQu6pZJ1ErWvkpQ, RTq8D5JqRWaTyFALEZIwkg, oK0gS38tQIa20YqftdLPgw], dist=MODULO}, downstreamNodes=[72yRyTOaQu6pZJ1ErWvkpQ, RTq8D5JqRWaTyFALEZIwkg, oK0gS38tQIa20YqftdLPgw], downstreamPhase=1, downstreamInputId=0}]
Leafs: []
target-sources: [3=>[2], 2=>[1], 1=>[0]]
original-error: UnsupportedOperationException: Can't handle Symbol [Reference: price]
```

**Steps to reproduce**:

1. Start cluster with crate 4.1.8
2. Prepare table
```
create table test (shape GEO_SHAPE);
```
3. Check SQL is successful
```
SELECT * FROM test WHERE MATCH (shape, 'POINT(1.2 1.3)')
```
4. Upgrade one node to crate 4.2.3. Await for a new node joined to cluster. 
5. Execute on upgraded(!) node:
```
SELECT * FROM test WHERE MATCH (shape, 'POINT(1.2 1.3)')
```
You will receive
```
SQLActionException[NotSerializableExceptionWrapper: unhandled_server_exception: java.lang.NullPointerException]
```

**Comment by seut** [comment]
@shm-dmitry Thank you a lot for reporting. We've fixed this bug by https://github.com/crate/crate/pull/10421. It will be published with the next 4.2.4 hotfix release.

**Comment by shm-dmitry** [comment]
@seut , thank you!
I tryed to run 4.2.4, but receive other errors from SQLs.
I think I need to use full-cluster-restart.

```
Caused by org.postgresql.util.PSQLException: ERROR: SQLParseException: Couldn't create executionContexts from
NodeOperations: [1=>NodeOp{ phase{id=1/distributed merge, nodes=[72yRyTOaQu6pZJ1ErWvkpQ, 5h_EYwpjQYy0My4Y_Z0cIQ, RTq8D5JqRWaTyFALEZIwkg, oK0gS38tQIa20YqftdLPgw], dist=BROADCAST}, downstreamNodes=[RTq8D5JqRWaTyFALEZIwkg], downstreamPhase=2, dow
nstreamInputId=0}, 0=>NodeOp{ phase{id=0/collect, nodes=[5h_EYwpjQYy0My4Y_Z0cIQ, 72yRyTOaQu6pZJ1ErWvkpQ, RTq8D5JqRWaTyFALEZIwkg, oK0gS38tQIa20YqftdLPgw], dist=MODULO}, downstreamNodes=[5h_EYwpjQYy0My4Y_Z0cIQ, 72yRyTOaQu6pZJ1ErWvkpQ, RTq8D5JqRW
aTyFALEZIwkg, oK0gS38tQIa20YqftdLPgw], downstreamPhase=1, downstreamInputId=0}]
Leafs: []
target-sources: [2=>[1], 1=>[0]]
original-error: UnsupportedOperationException: Unknown function: collect_set(INPUT(1)), no overload found for matching argument types: (text_array). Possible candidates: collect_set(char):array(char), collect_set(boolean):array(boolean), colle
ct_set(text):array(text), collect_set(ip):array(ip), collect_set(double precision):array(double precision), collect_set(real):array(real), collect_set(smallint):array(smallint), collect_set(integer):array(integer), collect_set(interval):array(
interval), collect_set(bigint):array(bigint), collect_set(timestamp with time zone):array(timestamp with time zone), collect_set(timestamp without time zone):array(timestamp without time zone)
```

**Comment by seut** [comment]
@shm-dmitry Ok seems like we have another issue. We'll have a look asap. Thank you again for reporting.

> I think I need to use full-cluster-restart.

To workaround this issue unfortunately yes.

**Comment by seut** [comment]
@shm-dmitry Sorry for the delay. I'm having problems to reproduce this error. Do you may know which query caused this?

**Comment by shm-dmitry** [comment]
@seut 
Sorry, I cant start mixed cluster for a long time..
I found broken SQL and created sample:

```
create table test (id int, id2 string);

select id, count(distinct(id))
from (
select id, id2 from test
) c
group by id;
```

**Comment by seut** [comment]
@shm-dmitry Ok thank you a lot, I've found the issue and working on a fix. Will keep you updated.

**Comment by seut** [comment]
@shm-dmitry We've fixed this issue, the fix will be part of the upcoming 4.2.5 hotfix release.
Thank you again for reporting.
